src/main/java/org/breedinginsight/services/geno/impl/GigwaGenoServiceImpl.java

```diff
@@ -0,0 +1,709 @@
+                                        .build();
+            } else {
+                return new GermplasmGenotype();
+            }
+        } catch (InstantiationException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean verifyProgramExists(BrAPIClient genoBrAPIClient, Program program) throws ApiException, InstantiationException {
+        ProgramsApi programsApi = brAPIEndpointProvider.get(genoBrAPIClient, ProgramsApi.class);
+        ApiResponse<BrAPIProgramListResponse> brAPIProgramListResponseApiResponse = programsApi.programsGet(new ProgramQueryParams().programDbId(program.getKey()));
+
+        return brAPIProgramListResponseApiResponse.getBody().getResult().getData().size() == 1;
+    }
+
+    private List<BrAPISample> fetchSamples(BrAPIClient genoBrAPIClient, Program program, List<BrAPIObservationUnit> observationUnits) throws ApiException, InstantiationException {
+        log.debug("fetching samples for OUs");
+        if(observationUnits.isEmpty()) {
+            log.debug("No OUs were supplied, returning");
+            return new ArrayList<>();
+        }
+
+        SamplesApi samplesApi = brAPIEndpointProvider.get(genoBrAPIClient, SamplesApi.class);
+
+        BrAPISampleSearchRequest sampleSearchRequest = new BrAPISampleSearchRequest();
+
+        sampleSearchRequest.setGermplasmDbIds(observationUnits.stream().map(ou -> program.getKey() + "§" + Utilities.removeProgramKeyAndUnknownAdditionalData(ou.getObservationUnitName(), program.getKey())).collect(Collectors.toList()));
+
+        return brAPIDAOUtil.search(samplesApi::searchSamplesPost, samplesApi::searchSamplesSearchResultsDbIdGet, sampleSearchRequest);
+    }
+
+    private List<BrAPIObservationUnit> fetchObservationUnits(BrAPIClient phenoBrAPIClient, BrAPIGermplasm germplasm) throws ApiException, InstantiationException {
+        ObservationUnitsApi observationUnitsApi = brAPIEndpointProvider.get(phenoBrAPIClient, ObservationUnitsApi.class);
+
+        BrAPIObservationUnitSearchRequest searchRequest = new BrAPIObservationUnitSearchRequest();
+        searchRequest.addGermplasmDbIdsItem(germplasm.getGermplasmDbId());
+
+        return brAPIDAOUtil.search(observationUnitsApi::searchObservationunitsPost, observationUnitsApi::searchObservationunitsSearchResultsDbIdGet, searchRequest);
+    }
+
+    private List<BrAPIObservationUnit> fetchObservationUnits(BrAPIClient phenoBrAPIClient, UUID experimentId) throws ApiException, DoesNotExistException, InstantiationException {
+        log.debug("fetching observationUnits for experiment: " + experimentId);
+        TrialsApi trialsApi = brAPIEndpointProvider.get(phenoBrAPIClient, TrialsApi.class);
+        ApiResponse<BrAPITrialListResponse> brAPITrialListResponseApiResponse = trialsApi.trialsGet(new TrialQueryParams().externalReferenceID(experimentId.toString())
```

Comment:
This will return all the trials on the breedbase server (or as many an the default pageSize will allow), not just the one asked specified.